src/Microsoft.ML.Api/ApiUtils.cs

```diff
@@ -81,21 +97,56 @@ private static OpCode GetAssignmentOpCode(Type t)
             return mb.CreateDelegate(typeof(Peek<TRow, TValue>));
         }
 
+        private static Delegate GeneratePeek<TOwn, TRow, TValue>(PropertyInfo propertyInfo, OpCode assignmentOpCode)
+        {
+            // REVIEW: It seems like we really should cache these, instead of generating them per cursor.
+            Type[] args = { typeof(TOwn), typeof(TRow), typeof(long), typeof(TValue).MakeByRefType() };
+            var mb = new DynamicMethod("Peek", null, args, typeof(TOwn), true);
+            var il = mb.GetILGenerator();
+
+            il.Emit(OpCodes.Ldarg_3);               // push arg3
+            il.Emit(OpCodes.Ldarg_1);               // push arg1
+            il.Emit(OpCodes.Call, propertyInfo.GetGetMethod());      // push [stack top].[propertyInfo]
```

Comment:
>Call [](start = 28, length = 4)

Should it be just call or callvirt?

Reply:
For F# records "call" was fine, but you're right that for C# get/set we may need callvirt

Reply:
We could just check the property, right?  If the property is virtual or abstract, then `callvirt`.  If it isn't virtual, then just `call`?

---
In reply to: [206686593](https://github.com/dotnet/machinelearning/pull/616#discussion_r206686593) [](ancestors = 206686593)